cfpb/consumerfinance.gov
# Restrict raw HTML to SVG icon use only

A long time ago, in a codebase not so far away, we created a global template function, `render_stream_child`, to render Wagtail stream field child blocks. We also had that block mark the rendered result as safe regardless of block type, and later we even unescaped HTML entities within those child blocks, enabling users to enter raw HTML in rich text fields. 

With the enormous condescension of posterity, and summarized so free of context, this series decisions sounds unfortunate. However, there remain some decent use-cases for entering HTML directly in the rich text editor that we don't have an answer for: adding our SVG icons. All other cases in rich text have been remediated though, and we're in the process of remediating the remaining handful of other fields.

This PR permits the SVG icons to remain in place where they are until we have a solution, but should 🤞🏻 eliminate the ability to add any other "raw HTML" anywhere else and have it rendered as HTML, and give us an avenue toward removing that capability altogether once we have an alternative approach for icons.

I've done this by adding a subclass of `RichTextField` that performs the `html.unescape()` that we were previously doing in `render_stream_child`. This moved that into the specific field that needs it, rather than on all stream field child blocks. Through our audit work we've found that the only current fields with SVG icon HTML in them are:

- `FullWidthText`'s `content` `RichTextBlock`.
- `Well`'s  `content` `RichTextBlock`.
- `TextIntroduction`'s `body` `RichTextBlock`.

So I have replaced their `RichTextBlock`s with `UnescapedRichTextBlock`s in place. This generates a migration, but because there's no change to the schema or data storage no data migration will be needed here, or (once SVG mitigation happens) to convert these back to `RichTextBlock`. 

Finally, our `render_stream_child` function predates Wagtail's own `include_block` template tag. Because we were relying on that function for rendering raw HTML entered in the CMS, we couldn't replace it with `include_block`. Now that this is no longer the case, I have removed `render_stream_child` and replaced it with `include_block`, removing one more bit of custom code we had that duplicated built-in Wagtail functionality.

There is some more detail in GHE in Design-and-Content-Team#428.

Until we complete the remaining non-SVG remediation this PR is on hold.

## How to test this PR

1. Pull the branch
2. Before we've remediated all non-SVG cases, check out Design-and-Content-Team#428 and see that they render the "raw HTML" inline and visible to the visitor now.
3. Use the `page_search` management command to get a list of current pages with "raw HTML" stored within `&lt;`/`&gt;` HTML entities in rich text fields: `python cfgov/manage.py page_search -s '&lt;[a-zA-Z]+((?!&gt;).)*&gt;'` 
4. Visit some of those pages and observe that they continue to render that raw HTML as expected.
5. Try to enter raw HTML into one of those fields that is not an SVG.

## Screenshots

If one attempts to enter an unsupported HTML element in an `UnescapedRichTextBlock` (i.e. any element that isn't `<svg>` and `<path>`): 
<img width="342" alt="image" src="https://github.com/cfpb/consumerfinance.gov/assets/10562538/dd39f904-aa23-4484-baf5-bf49583fcd4b">

## Checklist
